File: src/algebra2_logarithmic_equations_medium.py
"""
algebra2 - logarithmic_equations (medium)
Generated: 2026-02-22T04:32:40.054385
"""

from problem_utils import *

def generate():
    problem_type = choice(['basic_log', 'exponential_form', 'combining_logs', 'log_both_sides', 'two_step'])

    if problem_type == 'basic_log':
        # log_b(ax + c) = n where answer is integer
        base = choice([2, 3, 4, 5, 10])
        power = randint(1, 4)
        a_val = randint(1, 5)
        c_val = randint(0, 10)
        # x = (base^power - c) / a
        num = base**power - c_val
        if num > 0 and num % a_val == 0:
            x_val = num // a_val
        else:
            a_val = 1; c_val = randint(0, 5)
            x_val = base**power - c_val
            if x_val <= 0:
                c_val = 0; x_val = base**power
        ans = x_val

        if a_val == 1 and c_val == 0:
            inner = "x"
        elif a_val == 1:
            inner = f"x + {c_val}" if c_val > 0 else f"x"
        elif c_val == 0:
            inner = f"{a_val}x"
        else:
            inner = f"{a_val}x + {c_val}"

        question = f"\\log_{{{base}}}({inner}) = {power}"

        return problem(
            question=f"Solve for $x$: ${question}$",
            answer=ans,
            difficulty=(1300, 1450),
            topic="algebra2/logarithmic_equations",
            solution=steps(
                f"Convert to exponential form: ${inner} = {base}^{{{power}}} = {base**power}$",
                f"Solve: $x = {ans}$"
            )
        )

    elif problem_type == 'exponential_form':
        # Solve log(ax + b) = c for various bases
        base = choice([2, 3, 5, 10])
        a_val = randint(1, 4)
        c_val = randint(1, 3)
        # pick x_val then compute b
        x_val = randint(1, 10)
        b_val = base**c_val - a_val * x_val
        ans = x_val

        if base == 10:
            log_str = "\\log"
        else:
            log_str = f"\\log_{{{base}}}"

        if b_val >= 0:
            inner = f"{a_val}x + {b_val}" if a_val > 1 else f"x + {b_val}"
        else:
            inner = f"{a_val}x - {-b_val}" if a_val > 1 else f"x - {-b_val}"

        question = f"{log_str}({inner}) = {c_val}"

        solution_steps = [
            f"Convert to exponential form: ${inner} = {base}^{{{c_val}}} = {base**c_val}$",
        ]
        if a_val > 1:
            solution_steps.append(f"${a_val}x = {base**c_val - b_val}$")
        solution_steps.append(f"$x = {ans}$")

        return problem(
            question=f"Solve for $x$: ${question}$",
            answer=ans,
            difficulty=(1300, 1450),
            topic="algebra2/logarithmic_equations",
            solution=steps(*solution_steps)
        )

    elif problem_type == 'combining_logs':
        # log_b(x) + log_b(k) = c or log_b(x) - log_b(k) = c
        base = choice([2, 3, 5, 10])
        use_addition = choice([True, False])

        if use_addition:
            # log_b(x) + log_b(k) = c  => kx = base^c
            c_val = randint(1, 3)
            k_val = randint(2, 6)
            rhs = base**c_val
            if rhs % k_val == 0:
                ans = rhs // k_val
            else:
                k_val = 1; ans = rhs

            if base == 10:
                question = f"\\log(x) + \\log({k_val}) = {c_val}"
            else:
                question = f"\\log_{{{base}}}(x) + \\log_{{{base}}}({k_val}) = {c_val}"

            return problem(
                question=f"Solve for $x$: ${question}$",
                answer=ans,
                difficulty=(1400, 1550),
                topic="algebra2/logarithmic_equations",
                solution=steps(
                    f"Use product rule: $\\log_{{{base}}}({k_val}x) = {c_val}$",
                    f"Convert: ${k_val}x = {base}^{{{c_val}}} = {base**c_val}$",
                    f"$x = {ans}$"
                )
            )
        else:
            # log_b(x) - log_b(k) = c  => x/k = base^c
            c_val = randint(1, 3)
            k_val = randint(2, 8)
            ans = k_val * base**c_val

            if base == 10:
                question = f"\\log(x) - \\log({k_val}) = {c_val}"
            else:
                question = f"\\log_{{{base}}}(x) - \\log_{{{base}}}({k_val}) = {c_val}"

            return problem(
                question=f"Solve for $x$: ${question}$",
                answer=ans,
                difficulty=(1400, 1550),
                topic="algebra2/logarithmic_equations",
                solution=steps(
                    f"Use quotient rule: $\\log_{{{base}}}\\left(\\frac{{x}}{{{k_val}}}\\right) = {c_val}$",
                    f"Convert: $\\frac{{x}}{{{k_val}}} = {base}^{{{c_val}}} = {base**c_val}$",
                    f"$x = {k_val} \\cdot {base**c_val} = {ans}$"
                )
            )

    elif problem_type == 'log_both_sides':
        # b^(cx + d) = b^k => cx + d = k  (integer solution)
        base = choice([2, 3, 5])
        c_val = randint(1, 4)
        d_val = randint(-3, 3)
        k_val = randint(2, 5)
        # c*x + d = k  =>  x = (k - d)/c
        if (k_val - d_val) % c_val == 0:
            x_val = (k_val - d_val) // c_val
        else:
            c_val = 1; x_val = k_val - d_val
        ans = x_val
        rhs = base**k_val

        if d_val > 0:
            exp_str = f"{c_val}x + {d_val}" if c_val > 1 else f"x + {d_val}"
        elif d_val < 0:
            exp_str = f"{c_val}x - {-d_val}" if c_val > 1 else f"x - {-d_val}"
        else:
            exp_str = f"{c_val}x" if c_val > 1 else "x"

        question = f"{base}^{{{exp_str}}} = {rhs}"

        return problem(
            question=f"Solve for $x$: ${question}$",
            answer=ans,
            difficulty=(1350, 1500),
            topic="algebra2/logarithmic_equations",
            solution=steps(
                f"Take $\\log_{{{base}}}$ of both sides",
                f"${exp_str} = \\log_{{{base}}}({rhs}) = {k_val}$",
                f"Solve: $x = {ans}$"
            )
        )

    else:  # two_step
        # a*log_b(x) + c = d  =>  log_b(x) = (d-c)/a  =>  x = b^((d-c)/a)
        base = choice([2, 3, 5, 10])
        a_val = randint(2, 4)
        exponent = randint(1, 4)
        n_val = a_val * exponent
        c_val = randint(1, 8)
        d_val = n_val + c_val
        ans = base**exponent

        if base == 10:
            log_str = "\\log"
        else:
            log_str = f"\\log_{{{base}}}"

        question = f"{a_val}{log_str}(x) + {c_val} = {d_val}"

        return problem(
            question=f"Solve for $x$: ${question}$",
            answer=ans,
            difficulty=(1400, 1550),
            topic="algebra2/logarithmic_equations",
            solution=steps(
                f"Subtract ${c_val}$: ${a_val}{log_str}(x) = {n_val}$",
                f"Divide by ${a_val}$: ${log_str}(x) = {exponent}$",
                f"Convert: $x = {base}^{{{exponent}}} = {ans}$"
            )
        )

emit(generate())

File: src/algebra2_logarithmic_equations_medium.jl
# algebra2 - logarithmic_equations (medium)
# Generated: 2026-03-08T20:21:08.326002

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra2/logarithmic_equations")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Type 1: log_b(ax + c) = d, solve for x (ELO ~1200-1400)
        base = choice([2, 3, 4, 5, 10])
        d = randint(1, 4)
        target = base^d
        a = nonzero(-5, 5)
        c = randint(-20, 20)
        # Ensure ax + c = target has integer solution
        ans = (target - c) // a
        if denominator(ans) != 1
            ans = randint(-15, 15)
            c = target - a * ans
        else
            ans = numerator(ans)
        end
        
        lhs_arg = a*x + c
        
        problem(
            question="Solve for \\(x\\): \\(\\log_{$(base)}($(tex(lhs_arg))) = $(d)\\)",
            answer=ans,
            difficulty=(1200, 1400),
            solution=steps(
                sol("Given", "\\log_{$(base)}($(tex(lhs_arg))) = $(d)"),
                "Convert to exponential form: \\($(tex(lhs_arg)) = $(base)^{$(d)} = $(target)\\)",
                sol("Solve", a*x + c ~ target),
                sol("Answer", ans)
            ),
            time=90
        )
        
    elseif problem_type == 2
        # Type 2: log(x) + log(y) = k, word problem style (ELO ~1400-1600)
        base = choice([2, 10])
        k = randint(2, 5)
        product = base^k
        # Find two factors
        factors = []
        for i in 2:isqrt(product)
            if product % i == 0
                push!(factors, (i, product ÷ i))
            end
        end
        if isempty(factors)
            factors = [(1, product)]
        end
        factor_pair = choice(factors)
        a_val, b_val = factor_pair
        
        if base == 10
            log_notation = "\\log"
        else
            log_notation = "\\log_{$(base)}"
        end
        
        problem(
            question="Solve for \\(x\\): \\($(log_notation)(x) + $(log_notation)($(b_val)) = $(k)\\)",
            answer=a_val,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Given", "$(log_notation)(x) + $(log_notation)($(b_val)) = $(k)"),
                "Use logarithm product rule: \\($(log_notation)(x \\cdot $(b_val)) = $(k)\\)",
                "Convert to exponential form: \\(x \\cdot $(b_val) = $(base)^{$(k)} = $(product)\\)",
                sol("Solve for x", "x = \\frac{$(product)}{$(b_val)} = $(a_val)")
            ),
            time=120
        )
        
    elseif problem_type == 3
        # Type 3: log_b(x) - log_b(y) = k (ELO ~1400-1600)
        base = choice([2, 3, 5, 10])
        k = randint(1, 3)
        ratio = base^k
        denominator_val = nonzero(2, 12)
        numerator_val = ratio * denominator_val
        
        if base == 10
            log_notation = "\\log"
        else
            log_notation = "\\log_{$(base)}"
        end
        
        problem(
            question="Solve for \\(x\\): \\($(log_notation)(x) - $(log_notation)($(denominator_val)) = $(k)\\)",
            answer=numerator_val,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Given", "$(log_notation)(x) - $(log_notation)($(denominator_val)) = $(k)"),
                "Use logarithm quotient rule: \\($(log_notation)\\left(\\frac{x}{$(denominator_val)}\\right) = $(k)\\)",
                "Convert to exponential form: \\(\\frac{x}{$(denominator_val)} = $(base)^{$(k)} = $(ratio)\\)",
                sol("Solve for x", "x = $(ratio) \\cdot $(denominator_val) = $(numerator_val)")
            ),
            time=120
        )
        
    elseif problem_type == 4
        # Type 4: Exponential equation requiring logs (ELO ~1500-1700)
        base = choice([2, 3, 5, 7])
        exponent_coeff = nonzero(-3, 3)
        power = randint(2, 5)
        rhs = base^power
        # base^(exponent_coeff * x) = rhs
        ans = power // exponent_coeff
        if denominator(ans) != 1
            ans_display = "\\frac{$(numerator(ans))}{$(denominator(ans))}"
            ans_val = float(ans)
        else
            ans_display = "$(numerator(ans))"
            ans_val = numerator(ans)
        end
        
        exponent_display = exponent_coeff == 1 ? "x" : exponent_coeff == -1 ? "-x" : "$(exponent_coeff)x"
        
        problem(
            question="Solve for \\(x\\): \\($(base)^{$(exponent_display)} = $(rhs)\\)",
            answer=ans_val,
            difficulty=(1500, 1700),
            solution=steps(
                sol("Given", "$(base)^{$(exponent_display)} = $(rhs)"),
                "Rewrite RHS as power of $(base): \\($(rhs) = $(base)^{$(power)}\\)",
                "Equate exponents: \\($(exponent_display) = $(power)\\)",
                sol("Solve for x", "x = $(ans_display)")
            ),
            time=150
        )
        
    else
        # Type 5: log equation with quadratic substitution (ELO ~1600-1800)
        base = choice([2, 3, 10])
        # (log_b(x))^2 + a*log_b(x) + c = 0
        # Let u = log_b(x), solve u^2 + au + c = 0
        root1 = randint(1, 4)
        root2 = randint(1, 4)
        a_coeff = -(root1 + root2)
        c_coeff = root1 * root2
        
        x1 = base^root1
        x2 = base^root2
        
        if base == 10
            log_notation = "\\log"
        else
            log_notation = "\\log_{$(base)}"
        end
        
        middle_term = a_coeff == 0 ? "" : a_coeff > 0 ? " + $(a_coeff)$(log_notation)(x)" : " - $(abs(a_coeff))$(log_notation)(x)"
        constant_term = c_coeff == 0 ? "" : c_coeff > 0 ? " + $(c_coeff)" : " - $(abs(c_coeff))"
        
        problem(
            question="Solve for \\(x\\): \\(($(log_notation)(x))^2$(middle_term)$(constant_term) = 0\\)",
            answer=fmt_set(Set([x1, x2])),
            difficulty=(1600, 1800),
            answer_type="set",
            solution=steps(
                sol("Given", "($(log_notation)(x))^2$(middle_term)$(constant_term) = 0"),
                "Let \\(u = $(log_notation)(x)\\), then \\(u^2 + $(a_coeff)u + $(c_coeff) = 0\\)",
                "Factor: \\((u - $(root1))(u - $(root2)) = 0\\), so \\(u = $(root1)\\) or \\(u = $(root2)\\)",
                "Convert back: \\($(log_notation)(x) = $(root1) \\Rightarrow x = $(base)^{$(root1)} = $(x1)\\)",
                "And: \\($(log_notation)(x) = $(root2) \\Rightarrow x = $(base)^{$(root2)} = $(x2)\\)",
                sol("Solution set", "\\{$(x1), $(x2)\\}")
            ),
            time=180
        )
    end
end